src/renderer/components/PerCurrencySelectAccount/index.js

```diff
@@ -0,0 +1,218 @@
+import { withTranslation } from "react-i18next";
+import { connect } from "react-redux";
+import { createFilter } from "react-select";
+import { createStructuredSelector } from "reselect";
+import { shallowAccountsSelector } from "~/renderer/reducers/accounts";
+import Select from "~/renderer/components/Select";
+import type { CryptoCurrency, TokenCurrency } from "@ledgerhq/live-common/lib/types/currencies";
+import type { SubAccount } from "@ledgerhq/live-common/lib/types/account";
+import { MenuOption } from "~/renderer/components/PerCurrencySelectAccount/Option";
+import { BigNumber } from "bignumber.js";
+
+const mapStateToProps = createStructuredSelector({
+  accounts: shallowAccountsSelector,
+});
+
+type Option = {
+  matched: "boolean",
+  account: Account,
+  subAccount: SubAccount,
+};
+const getOptionValue = option => {
+  return option.account ? option.account.id : null;
+};
+
+const defaultFilter = createFilter({
+  stringify: ({ data: account }) => {
+    const currency = getAccountCurrency(account);
+    const name = getAccountName(account);
+    return `${currency.ticker}|${currency.name}|${name}`;
+  },
+});
+const filterOption = o => (candidate, input) => {
+  const selfMatches = defaultFilter(candidate, input);
+  if (selfMatches) return [selfMatches, true];
+  return [false, false];
+};
+
+const AccountOption = React.memo(function AccountOption({
+  account,
+  subAccount,
+  currency,
+  isValue,
+}: {
+  account: Account,
+  subAccount: SubAccount,
+  currency: CryptoCurrency | TokenCurrency,
+  isValue?: boolean,
+}) {
+  return <MenuOption isValue={isValue} account={account} subAccount={subAccount} />;
+});
+
+type OwnProps = {
+  currency: TokenCurrency | CryptoCurrency,
+  mandatoryTokens?: boolean,
+  value: ?AccountLike,
+  onChange: (account: ?AccountLike, tokenAccount: ?Account) => void,
+};
+
+type Props = OwnProps & {
+  accounts: Account[],
+  t: TFunction,
+  onChange: (account: ?AccountLike, tokenAccount: ?Account) => void, // I have to add it twice else flow is unhappy, don't ask why
+};
+
+const getAccountsByCurrency = (accounts: Account[], currency: TokenCurrency | CryptoCurrency) =>
+  accounts.filter((acc: Account) => acc.currency.id === currency.id);
+
+// TODO: token account generation logic in live-common
+const generateTokenAccount = (account: Account, currency: TokenCurrency) => ({
+  type: "TokenAccount",
+  id: account.id + "+" + currency.contractAddress,
+  parentId: account.id,
+  token: currency,
+  balance: BigNumber(0),
+  operationsCount: 0,
+  creationDate: new Date(),
+  operations: [],
+  pendingOperations: [],
+  starred: false,
+});
+
+const RawSelectAccount = ({
+  accounts,
+  onChange,
+  t,
+  currency,
+  mandatoryTokens,
+  ...props
+}: Props) => {
+  const availableAccounts = useMemo(() => {
+    if (currency.type === "TokenCurrency") {
+      const mainAccounts = getAccountsByCurrency(accounts, currency.parentCurrency);
+      return mainAccounts.map((acc: Account) => {
```

Comment:
This should be using `accountWithMandatoryTokens` instead, just provide the currency and map the accounts, something like `mainAccounts.map(a=>accountWithMandatoryTokens(a, [currency])` would return all accounts with the needed subaccounts without destroying existing ones

Reply:
My understanding is that accountWithMandatory would indeed simplify some code here and as you will have to do the same on mobile is better as shared code.

But at the same time this seems to be implemented in a different paradigm where values are actually tuples with sub accounts? I'm ok with that PR solution that seems better to keep the account reference and avoid bad surprise.
However that generateTokenAccount must be moved to live-common. Please provide a PR to make it there.

Reply:
Yep that's exactly how I've seen it. I find it less convenient to mutate accounts. Account generation was marked for factorisation in live common already. accountWithMandatoryToken should use this func too !